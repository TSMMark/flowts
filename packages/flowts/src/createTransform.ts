import * as babel from '@babel/core';
import * as prettier from 'prettier';
import recastPlugin from '@flowts/babel-plugin-recast';
import { PluginItem } from '@babel/core';
import { Options } from 'prettier';

/**
 * Helper utility creating function transforming the code given plugins array
 * @param plugins
 * @param prettierConfig
 * @param recast
 */
export function createTransform(
  plugins: PluginItem[],
  { prettierConfig = {} as Options, recast = true } = {}
) {
  const transformPlugins: PluginItem[] = [];
  if (recast) {
    transformPlugins.push(recastPlugin);
  }

  return function transform(code: string) {
    const transformResult = babel.transformSync(code, {
      compact: false,
      babelrc: false,
      configFile: false,
      plugins: [...transformPlugins, ...plugins],
    });

    if (transformResult === null) {
      throw new Error(
        'babel.transformSync returned null, probably configuration issue'
      );
    }
    prettierConfig.parser = 'babel-ts';
    return prettier.format(transformResult.code as string, {
      ...prettierConfig,
      // avoid changing template strings
      embeddedLanguageFormatting: 'off',
    });
  };
}
