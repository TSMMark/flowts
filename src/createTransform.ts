import * as babel from '@babel/core';
import * as prettier from 'prettier';
import recastPlugin from './recastPlugin';
import { PluginItem } from '@babel/core';
import { Options } from 'prettier';

/**
 * Helper utility creating function transforming the code given plugins array
 * @param plugins
 * @param prettierConfig
 * @param recast
 */
export function createTransform(
  plugins: PluginItem[],
  { prettierConfig = {} as Options, recast = true } = {}
) {
  const transformPlugins: PluginItem[] = [];
  if (recast) {
    transformPlugins.push(recastPlugin);
  }

  return function transform(code: string) {
    const transformResult = babel.transformSync(code, {
      babelrc: false,
      plugins: [...transformPlugins, ...plugins],
    });

    if (transformResult === null) {
      throw new Error(
        'babel.transformSync returned null, probably configuration issue'
      );
    }
    prettierConfig.parser = 'typescript';
    return prettier.format(transformResult.code as string, prettierConfig);
  };
}
