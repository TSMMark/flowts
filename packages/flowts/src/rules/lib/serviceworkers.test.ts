import plugin from '../../tsTypesPlugin';
import { createTransform } from '../../createTransform';
const transform = createTransform([plugin]);

describe('lib/serviceworkers.js', () => {
  describe('globals', () => {
    describe('WindowClient', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = WindowClient;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new WindowClient();
            
            class A1 extends WindowClient {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('Client', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = Client;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new Client();
            
            class A1 extends Client {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('ExtendableEvent', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = ExtendableEvent;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new ExtendableEvent();
            
            class A1 extends ExtendableEvent {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('InstallEvent', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = InstallEvent;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new InstallEvent();
            
            class A1 extends InstallEvent {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('FetchEvent', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = FetchEvent;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new FetchEvent();
            
            class A1 extends FetchEvent {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('Clients', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = Clients;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new Clients();
            
            class A1 extends Clients {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('ServiceWorker', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = ServiceWorker;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new ServiceWorker();
            
            class A1 extends ServiceWorker {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('NavigationPreloadState', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = NavigationPreloadState;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new NavigationPreloadState();
            
            class A1 extends NavigationPreloadState {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('NavigationPreloadManager', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = NavigationPreloadManager;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new NavigationPreloadManager();
            
            class A1 extends NavigationPreloadManager {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('PushSubscriptionJSON', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = PushSubscriptionJSON;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new PushSubscriptionJSON();
            
            class A1 extends PushSubscriptionJSON {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('PushSubscription', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = PushSubscription;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new PushSubscription();
            
            class A1 extends PushSubscription {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('PushManager', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = PushManager;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new PushManager();
            
            class A1 extends PushManager {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('ServiceWorkerRegistration', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = ServiceWorkerRegistration;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new ServiceWorkerRegistration();
            
            class A1 extends ServiceWorkerRegistration {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('ServiceWorkerContainer', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = ServiceWorkerContainer;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new ServiceWorkerContainer();
            
            class A1 extends ServiceWorkerContainer {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('ServiceWorkerMessageEvent', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = ServiceWorkerMessageEvent;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new ServiceWorkerMessageEvent();
            
            class A1 extends ServiceWorkerMessageEvent {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('ExtendableMessageEvent', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = ExtendableMessageEvent;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new ExtendableMessageEvent();
            
            class A1 extends ExtendableMessageEvent {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('Cache', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = Cache;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new Cache();
            
            class A1 extends Cache {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('CacheStorage', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = CacheStorage;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new CacheStorage();
            
            class A1 extends CacheStorage {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('clients', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = clients;
          `)
        ).toMatchSnapshot();
      });
    });

    describe('caches', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = caches;
          `)
        ).toMatchSnapshot();
      });
    });

    describe('registration', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = registration;
          `)
        ).toMatchSnapshot();
      });
    });

    describe('skipWaiting', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = skipWaiting;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new skipWaiting();
            
            class A1 extends skipWaiting {};
          `)
        ).toMatchSnapshot();
      });

      test('generated - callable', () => {
        expect(
          transform(`
            skipWaiting();
            
          `)
        ).toMatchSnapshot();
      });
    });

    describe('onactivate', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = onactivate;
          `)
        ).toMatchSnapshot();
      });
    });

    describe('oninstall', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = oninstall;
          `)
        ).toMatchSnapshot();
      });
    });

    describe('onfetch', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = onfetch;
          `)
        ).toMatchSnapshot();
      });
    });

    describe('onforeignfetch', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = onforeignfetch;
          `)
        ).toMatchSnapshot();
      });
    });

    describe('onmessage', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = onmessage;
          `)
        ).toMatchSnapshot();
      });
    });
  });

  describe('modules', () => {
    test('has no modules', () => {});
  });
});
