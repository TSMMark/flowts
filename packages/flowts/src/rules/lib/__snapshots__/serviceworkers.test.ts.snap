// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/serviceworkers.js globals Cache generated - class 1`] = `
"new Cache();

class A1 extends Cache {}
"
`;

exports[`lib/serviceworkers.js globals Cache variable 1`] = `
"var a = Cache;
"
`;

exports[`lib/serviceworkers.js globals CacheStorage generated - class 1`] = `
"new CacheStorage();

class A1 extends CacheStorage {}
"
`;

exports[`lib/serviceworkers.js globals CacheStorage variable 1`] = `
"var a = CacheStorage;
"
`;

exports[`lib/serviceworkers.js globals Client generated - class 1`] = `
"new Client();

class A1 extends Client {}
"
`;

exports[`lib/serviceworkers.js globals Client variable 1`] = `
"var a = Client;
"
`;

exports[`lib/serviceworkers.js globals Clients generated - class 1`] = `
"new Clients();

class A1 extends Clients {}
"
`;

exports[`lib/serviceworkers.js globals Clients variable 1`] = `
"var a = Clients;
"
`;

exports[`lib/serviceworkers.js globals ExtendableEvent generated - class 1`] = `
"new ExtendableEvent();

class A1 extends ExtendableEvent {}
"
`;

exports[`lib/serviceworkers.js globals ExtendableEvent variable 1`] = `
"var a = ExtendableEvent;
"
`;

exports[`lib/serviceworkers.js globals ExtendableMessageEvent generated - class 1`] = `
"new ExtendableMessageEvent();

class A1 extends ExtendableMessageEvent {}
"
`;

exports[`lib/serviceworkers.js globals ExtendableMessageEvent variable 1`] = `
"var a = ExtendableMessageEvent;
"
`;

exports[`lib/serviceworkers.js globals FetchEvent generated - class 1`] = `
"new FetchEvent();

class A1 extends FetchEvent {}
"
`;

exports[`lib/serviceworkers.js globals FetchEvent variable 1`] = `
"var a = FetchEvent;
"
`;

exports[`lib/serviceworkers.js globals InstallEvent generated - class 1`] = `
"new InstallEvent();

class A1 extends InstallEvent {}
"
`;

exports[`lib/serviceworkers.js globals InstallEvent variable 1`] = `
"var a = InstallEvent;
"
`;

exports[`lib/serviceworkers.js globals NavigationPreloadManager generated - class 1`] = `
"new NavigationPreloadManager();

class A1 extends NavigationPreloadManager {}
"
`;

exports[`lib/serviceworkers.js globals NavigationPreloadManager variable 1`] = `
"var a = NavigationPreloadManager;
"
`;

exports[`lib/serviceworkers.js globals NavigationPreloadState generated - class 1`] = `
"new NavigationPreloadState();

class A1 extends NavigationPreloadState {}
"
`;

exports[`lib/serviceworkers.js globals NavigationPreloadState variable 1`] = `
"var a = NavigationPreloadState;
"
`;

exports[`lib/serviceworkers.js globals PushManager generated - class 1`] = `
"new PushManager();

class A1 extends PushManager {}
"
`;

exports[`lib/serviceworkers.js globals PushManager variable 1`] = `
"var a = PushManager;
"
`;

exports[`lib/serviceworkers.js globals PushSubscription generated - class 1`] = `
"new PushSubscription();

class A1 extends PushSubscription {}
"
`;

exports[`lib/serviceworkers.js globals PushSubscription variable 1`] = `
"var a = PushSubscription;
"
`;

exports[`lib/serviceworkers.js globals PushSubscriptionJSON generated - class 1`] = `
"new PushSubscriptionJSON();

class A1 extends PushSubscriptionJSON {}
"
`;

exports[`lib/serviceworkers.js globals PushSubscriptionJSON variable 1`] = `
"var a = PushSubscriptionJSON;
"
`;

exports[`lib/serviceworkers.js globals ServiceWorker generated - class 1`] = `
"new ServiceWorker();

class A1 extends ServiceWorker {}
"
`;

exports[`lib/serviceworkers.js globals ServiceWorker variable 1`] = `
"var a = ServiceWorker;
"
`;

exports[`lib/serviceworkers.js globals ServiceWorkerContainer generated - class 1`] = `
"new ServiceWorkerContainer();

class A1 extends ServiceWorkerContainer {}
"
`;

exports[`lib/serviceworkers.js globals ServiceWorkerContainer variable 1`] = `
"var a = ServiceWorkerContainer;
"
`;

exports[`lib/serviceworkers.js globals ServiceWorkerMessageEvent generated - class 1`] = `
"new ServiceWorkerMessageEvent();

class A1 extends ServiceWorkerMessageEvent {}
"
`;

exports[`lib/serviceworkers.js globals ServiceWorkerMessageEvent variable 1`] = `
"var a = ServiceWorkerMessageEvent;
"
`;

exports[`lib/serviceworkers.js globals ServiceWorkerRegistration generated - class 1`] = `
"new ServiceWorkerRegistration();

class A1 extends ServiceWorkerRegistration {}
"
`;

exports[`lib/serviceworkers.js globals ServiceWorkerRegistration variable 1`] = `
"var a = ServiceWorkerRegistration;
"
`;

exports[`lib/serviceworkers.js globals WindowClient generated - class 1`] = `
"new WindowClient();

class A1 extends WindowClient {}
"
`;

exports[`lib/serviceworkers.js globals WindowClient variable 1`] = `
"var a = WindowClient;
"
`;

exports[`lib/serviceworkers.js globals caches variable 1`] = `
"var a = caches;
"
`;

exports[`lib/serviceworkers.js globals clients variable 1`] = `
"var a = clients;
"
`;

exports[`lib/serviceworkers.js globals onactivate variable 1`] = `
"var a = onactivate;
"
`;

exports[`lib/serviceworkers.js globals onfetch variable 1`] = `
"var a = onfetch;
"
`;

exports[`lib/serviceworkers.js globals onforeignfetch variable 1`] = `
"var a = onforeignfetch;
"
`;

exports[`lib/serviceworkers.js globals oninstall variable 1`] = `
"var a = oninstall;
"
`;

exports[`lib/serviceworkers.js globals onmessage variable 1`] = `
"var a = onmessage;
"
`;

exports[`lib/serviceworkers.js globals registration variable 1`] = `
"var a = registration;
"
`;

exports[`lib/serviceworkers.js globals skipWaiting generated - callable 1`] = `
"skipWaiting();
"
`;

exports[`lib/serviceworkers.js globals skipWaiting generated - class 1`] = `
"new skipWaiting();

class A1 extends skipWaiting {}
"
`;

exports[`lib/serviceworkers.js globals skipWaiting variable 1`] = `
"var a = skipWaiting;
"
`;
